frame/data-feed/src/lib.rs

```diff
@@ -0,0 +1,704 @@
+		SetUrl(StorageKey, Vec<u8>),
+		/// New data stored on blockchain
+		FeedData(AccountId, StorageKey, DataType),
+		/// Set calculated data
+		SetData(StorageKey, DataType),
+	}
+);
+
+// Errors inform users that something went wrong.
+decl_error! {
+	pub enum Error for Module<T: Trait> {
+		/// If the storage key already registered
+		ExistedKey,
+		/// You can not put data under the key that is not registered
+		InvalidKey,
+		/// If the data type does not match the one specified in Infos
+		InvalidValue,
+		/// user is not allowed to feed data onto the blockchain
+		NotAllowed,
+		/// provider count of key count exceed limit
+		ExceedLimit,
+	}
+}
+
+decl_storage! {
+	trait Store for Module<T: Trait> as DataFeed {
+		/// A set of storage keys
+		/// A storage key in this set refers to a storage value in other pallets,
+		pub ActiveParamTypes get(fn all_keys): Vec<StorageKey>;
+
+		/// A set of accounts
+		/// Any one can submit data on chain, while only the data submitted by
+		/// the provider in this set is valid, and will be calculated later.
+		pub ActiveProviders get(fn all_providers): Vec<T::AccountId>;
+
+		/// Data Attributes.
+		/// It defines the type of the data, how to extract the data we want
+		/// from a json blob
+		pub Infos get(fn infos): map hasher(twox_64_concat) StorageKey => Option<Info<T::BlockNumber>>;
+
+		/// permissioned URL that could be used to fetch data
+		pub Url get(fn url): map hasher(twox_64_concat) StorageKey => Option<Vec<u8>>;
+
+		pub DataFeeds get(fn feeded_data): double_map hasher(blake2_128_concat) StorageKey,
+			hasher(blake2_128_concat) T::AccountId => Option<[DataType; RING_BUF_LEN]>;
+	}
+}
+
+impl<T: Trait> Contains<T::AccountId> for Module<T> {
+	fn sorted_members() -> Vec<T::AccountId> {
+		Self::all_providers()
+	}
+}
+
+decl_module! {
+	pub struct Module<T: Trait> for enum Call where origin: T::Origin {
+		type Error = Error<T>;
+		fn deposit_event() = default;
+
+		/// Register a storage key under which the value can be modified
+		/// by this data feed pallet with some rules
+		#[weight = 0]
+		pub fn register_storage_key(origin, key: StorageKey, info: Info<T::BlockNumber>) -> DispatchResult {
+			T::DispatchOrigin::try_origin(origin).map(|_| ()).or_else(ensure_root)?;
+			// parse origin
+			ActiveParamTypes::try_mutate::<_, DispatchError, _>(|v| {
+				if v.contains(&key) {
+					Err(Error::<T>::ExistedKey)?;
+				}
+				if v.len() + 1 > MAX_KEY_LEN {
+					Err(Error::<T>::ExceedLimit)?
+				}
+
+				v.push(key.clone());
+				Ok(())
+			})?;
+			Infos::<T>::insert(&key, info);
+			Self::deposit_event(RawEvent::RegisterStorageKey(key));
+			Ok(())
+		}
+
+		/// remove the storage key from the limited set so the data feed pallet no longer
+		/// can change the corresponding value afterwards.
+		#[weight = 0]
+		pub fn remove_storage_key(origin, key: StorageKey) -> DispatchResult {
+			T::DispatchOrigin::try_origin(origin).map(|_| ()).or_else(ensure_root)?;
```

Comment:
```suggestion
			T::DispatchOrigin::try_origin(origin)?;
```

You dont need to add `ensure_root` here. This can be configured by the runtime developer when configuring the `DispatchOrigin`

Reply:
fixed